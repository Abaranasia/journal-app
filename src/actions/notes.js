import { db } from "../firebase/firebaseConfig";
import { types } from "../types/types";

import { loadNotes } from "../helpers/loadNotes";
import Swal from "sweetalert2";


export const startNewNote = () => { // Acción para crear una nueva nota
    return async (dispatch, getState) => { //comunicación asíncrona con Thunk 

        const uid = getState().auth.uid;
        //console.log(uid)

        const newNote = {
            title: 'Otra nueva nota',
            body: 'jujijaje',
            date: new Date().getTime()
        }

        const docRef = await db.collection(`${uid}/journal/notes`).add(newNote); // referencia al documento en Firestore donde añadir una nueva nota para este usuario
        // console.log(docRef)
        dispatch(activeNote(docRef.id, newNote))
    }
}

export const activeNote = (id, note) => ({
    type: types.notesActive,
    payload: {
        id,
        ...note
    }

})

export const startLoadingNotes = (uid) => {
    return async (dispatch) => { // acción asíncrona porque depende del backend

        const notes = await loadNotes(uid); // obtenemos todas las notas del usuario identificado y las enviaremos al store
        dispatch(setNotes(notes)) //hacemos el dispatch para enviar el listado de notas al store mediante la acción setNotes
    }
}


export const setNotes = (notes) => ({ // Acción para enviar al store las notas recibidas de Firebase
    type: types.notesLoad,
    payload: notes
})

export const startSaveNote = (note) => { // Acción para guardar una nota en la BBDD
    // necesitaremos el middleware Thunk para poder comunicarnos con Firebase
    return async (dispatch, getState) => { //Thunk, que es asíncrono, devuelve el dispatch y el uid del usuario

        const { uid } = getState().auth;

        if (!note.url) {
            delete note.url; //Para evitar que Firebase pete si le llega un undefined
        }

        const noteToFirestore = { ...note };
        delete noteToFirestore.id; // Hacemos esto para extraer la id de note, dado que no la necesitamos en la inserción

        await db.doc(`${uid}/journal/notes/${note.id}`).update(noteToFirestore); // Podemos enviar noteToFirestore o note, pro la primera no lleva id, ojo

        dispatch(refreshNote(note.id, noteToFirestore));
        Swal.fire('Saved!', note.title, 'success');
    }
}

export const refreshNote = (id, note) => ({ // Acción que permite refrescar una nota, en caso de que haya cambiado, por ejemplo, tras guardarla
    type: types.notesUpdated,
    payload: {
        id,
        note: {  // Tenemos que pasarlo así porque este note nos llega sin id al enviar noteToFirestore, asi que lo volvemos a poner
            id, //Sin este este id, jornalEntry nos da un error porque debe dibujar un elemento sin key
            ...note
        }

    }
})

