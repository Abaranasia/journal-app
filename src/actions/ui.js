import { types } from '../types/types';

/** Este mÃ³dulo define los actions en referencia a los error del sistema */

export const setError = (err) => ({
    type: types.uiSetError,
    payload: err
})

export const removeError = () => ({
    type: types.uiRemoveError
})

export const startLoading = () => ({
    type: types.uiStartLoading,

})

export const finishLoading = () => ({
    type: types.uiFinishLoading,

})