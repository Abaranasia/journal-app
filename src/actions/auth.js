import { firebase, googleAuthProvider } from "../firebase/firebaseConfig";
import Swal from 'sweetalert2';

import { types } from "../types/types";
import { finishLoading, startLoading } from "./ui";
import { notesLogout } from "./notes";


/** Este módulo define los actions para el proceso de auth.
 *  Aquí incluiríamos todos los métodos de acceso necesarios: email, google, twitter, etc.
 */


export const startLoginEmailPassword = (email, password) => {
    // middleware para login con email que gestiona el action de login

    return (dispatch) => { //devuelve un objeto llamado dispatch con el tipo y el payload

        dispatch(startLoading()); // Empieza el procedimiento de login

        
        return firebase.auth().signInWithEmailAndPassword(email, password)
            .then(({ user }) => {
                dispatch(login(user.uid, user.displayName));
                dispatch(finishLoading()); // Finaliza el procedimiento de login 
            })
            .catch(e => {
                console.log('Hubo un error en el login:', e);
                dispatch(finishLoading()); // Finaliza el procedimiento de login 
                Swal.fire('Error', e.message, 'error')
            });
/*         setTimeout(() => {
            dispatch(login(123, 'Pedro'))
        }, 3500); */

    }
}

export const startRegisterWithEmailPassword = (email, password, name) => {
    // dado que es ua tarea asíncrona, necesito retornar un callback
    return (dispatch) => {
        firebase.auth().createUserWithEmailAndPassword(email, password)
            .then(async ({ user }) => {
                await user.updateProfile({ displayName: name }); // hacemos esto porque el registro en Firebase no lo hace solo
                console.log(user);

                dispatch(
                    login(user.uid, user.displayName)
                )
            })
            .catch(e => { console.log('Hubo un error al registrarse:', e) });
    }
}

export const startGoogleLogin = () => {
    // middleware para login con usuario de Google que gestiona el action de login

    return (dispatch) => {
    //devuelve una función asíncrona que recibe un dispatch con el tipo y el payload que deben ser enviados

        firebase.auth().signInWithPopup(googleAuthProvider)
            .then(({ user }) => {
                //console.log(user.uid, user.displayName);
                dispatch(
                    login(user.uid, user.displayName)
                )
            });
    }
}


export const login = (uid, displayName) => {
    //action para login: devuelve un tipo (login) y un payload con uid y nombre

    return {
        type: types.login,
        payload: {
            uid, displayName
        }
    }
}

export const startLogout = () => {
    return async (dispatch) => {
        await firebase.auth().signOut();

        dispatch(logout())
        dispatch(notesLogout())
    }
}
export const logout = (params) => ({
    type: types.logout
})
