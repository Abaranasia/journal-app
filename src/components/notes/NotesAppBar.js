import React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { startSaveNote, startUploading } from '../../actions/notes';

export const NotesAppBar = () => {

    const dispatch = useDispatch();
    const { active } = useSelector(state => state.notes);


    const handleSave = () => { // guarda una nota
        dispatch(startSaveNote(active))
    }

    const handlePictureClick = () => { //redirige el evento del envío de imagen
        document.querySelector('#fileSelector').click()
    }


    const handleFileChange = (e) => { //gestiona el envío de la imagen
        const file = (e.target.files[0]);
        console.log('FILE: ', file)
        if (file) { dispatch(startUploading(file)) }

    }
    return (
        <div className="notes__appbar">
            <span> 25 de septiembre de 2021</span>

            <input
                id="fileSelector"
                type="file"
                name="file"
                style={{ display: 'none' }}
                onChange={handleFileChange}
            />

            <div>
                <button className="btn" onClick={handlePictureClick}>
                    Picture
                </button>
                <button className="btn" onClick={handleSave}>
                    Save
                </button>
            </div>

        </div>
    )
}
