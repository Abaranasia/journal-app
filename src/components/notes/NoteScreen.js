import React, { useEffect, useRef } from 'react';
import { NotesAppBar } from './NotesAppBar';
import { useForm } from '../../hooks/useForm';

import { useDispatch, useSelector } from 'react-redux';
import { activeNote, startDeleting } from '../../actions/notes';


export const NoteScreen = () => {

    const dispatch = useDispatch();

    const { active: actvNote } = useSelector(state => state.notes); //obtiene la info de la nota activa
    const [formValues, handleInputChange, reset] = useForm(actvNote);
    const { body, title, id } = formValues;

    const activeId = useRef(actvNote.id); // Almacena una variable mutable que no va a redibujar todo el componente si cambia


    useEffect(() => { //Esto se disparará solo cuando la id de la nota activa cambie

        if (actvNote.id !== activeId.current) { // Si la nota activa es diferente de la actual, necesitamos resetear el form
            reset(actvNote);
            activeId.current = actvNote.id // Actualizamos y ahora la activa será la actual
        }

    }, [actvNote, reset])


    useEffect(() => {
        //console.log(formValues)
        dispatch(activeNote(formValues.id, { ...formValues })); //despachamos la acción (no confundir con actvNote) y le enviamos el contenido del form

    }, [formValues, dispatch])

    const handleDelete = () => { // Borra una nota
        dispatch(startDeleting(id)) //pasamos el id obtenido desde formValues
    }

    return (
        <div className="notes__main-content">

            <NotesAppBar />

            <div className="notes__content">
                <input
                    type="text"
                    placeholder="Título sorprendente"
                    className="notes__title-input"
                    autoComplete="off"
                    name="title"
                    value={title}
                    onChange={handleInputChange}
                />
                <textarea
                    placeholder="Cuénta qué pasó hoy"
                    className="notes__textarea"
                    name="body"
                    value={body}
                    onChange={handleInputChange}
                >
                </textarea>

                {
                    (actvNote.url) &&
                    (<div className="notes__image">
                    <img
                            src={actvNote.url}
                        alt="día"
                    />
                    </div>)
                }
            </div>

            <button className="btn btn-danger" onClick={handleDelete}>
                Delete
            </button>
        </div>
    )
}
