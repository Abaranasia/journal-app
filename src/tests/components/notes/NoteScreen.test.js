import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';

import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import '@testing-library/jest-dom';
//import { firebase } from '../../firebase/firebaseConfig';

import { NoteScreen } from '../../../components/notes/NoteScreen';
import { activeNote } from '../../../actions/notes';


//import { getAuth } from 'firebase/auth';
//const auth = getAuth();

//Evita el crasheo de tipo: INTERNAL ASSERTION FAILED: Expected a class definition en  firebase.auth()


jest.mock('../../../actions/notes', () => ({
    // Mockeamos esta función porque ya está probada y no necesitamos volver a hacerlo aquí
    activeNote: jest.fn(),
})
);


const middlewares = [thunk];
const mockStore = configureStore(middlewares);
// Para evaluar estas pruebas vamos a necesitar que nos ayude el store. Para ello empleamos mockStore, 
// MockStore nos permitirá averiguarlas acciones que lo han disparado, con qué argumentos, etc.

const initState = { //Debe tener el estado inicial, el que nos encontraremos antes de hacer login 
    auth: {
        udi: "1",
        name: "Ran"
    },
    ui: {
        loading: false,
        msgError: null
    },
    notes: {
        active: {
            id: 1234,
            title: "test active note",
            body: "test body",

        },
        notes: []
    }
}

let store = mockStore(initState);
// Nuestro store fictício debe definir aquí el estado inicial en el que necesitamos que esté para estos test
store.dispatch = jest.fn(); // Mokeamos el dispatch porque necesitamos evaluarlo aquí

const wrapper = mount(
    <Provider store={store}> {/* lo necesitamos para fingir el context del usuario */}
        <NoteScreen />
    </Provider>
)


describe('NoteScreen tests', () => {


    test('should render the component properly', () => {
        expect(wrapper).toMatchSnapshot()
    });


    test('should run activeNote', () => {
        wrapper.find('input[name="title"]').simulate('change', {
            target: {
                name: 'title',
                value: 'Hola de nuevo'
            }
        });

        expect(activeNote).toHaveBeenLastCalledWith(
            // Usamos toHaveBeenLastCalledWith porque en su primera ejecución dará un resultado distinto, el inicial, y peta
            1234,
            {
                body: 'test body',
                title: 'Hola de nuevo',
                id: 1234
            }
        );

    });
});