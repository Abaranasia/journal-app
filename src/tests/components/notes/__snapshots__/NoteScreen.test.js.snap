// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NoteScreen tests should render the component properly 1`] = `
Array [
  " ",
  <div
    className="notes__main-content"
  >
    <div
      className="notes__appbar"
    >
      <span>
         25 de septiembre de 2021
      </span>
      <input
        id="fileSelector"
        name="file"
        onChange={[Function]}
        style={
          Object {
            "display": "none",
          }
        }
        type="file"
      />
      <div>
        <button
          className="btn"
          onClick={[Function]}
        >
          Picture
        </button>
        <button
          className="btn"
          onClick={[Function]}
        >
          Save
        </button>
      </div>
    </div>
    <div
      className="notes__content"
    >
      <input
        autoComplete="off"
        className="notes__title-input"
        name="title"
        onChange={[Function]}
        placeholder="Título sorprendente"
        type="text"
        value="test active note"
      />
      <textarea
        className="notes__textarea"
        name="body"
        onChange={[Function]}
        placeholder="Cuénta qué pasó hoy"
        value="test body"
      />
    </div>
    <button
      className="btn btn-danger"
      onClick={[Function]}
    >
      Delete
    </button>
  </div>,
]
`;
